### PR TITLE
Improve DisplayList bounds calculations

### PR DESCRIPTION
**Pull request by flar**
These changes are to help improve our bounds calculations - and thus our cache memory usage - for the DisplayList.

The effort was triggered by the regression documented in https://github.com/flutter/flutter/issues/88745 which resulted in disabling the DisplayList mechanism by default in the product. Rerunning the tests confirms that this PR fixes https://github.com/flutter/flutter/issues/88745

Testing locally with the Flutter Gallery transition benchmarks we could see that our cache usage was higher with DisplayList and we've taken a number of steps to reduce that usage back to the baseline, including this PR and https://github.com/flutter/engine/pull/28158 and https://github.com/flutter/engine/pull/28287.

Improving our bounds resulted in a whole new algorithm for the `DisplayListBoundsCalculator` class which now does all of its own work except for a few cases where we deal with opaque Skia objects which must be trusted as sources or modifiers of rendering bounds (SkTextBlob, ShadowUtils, SkImageFilter, SkMaskFilter, SkPathEffect). Some reasons for introducing the new architecture were:

- Avoiding having to translate the DisplayList data into an SkPaint to compute the bounds
- Avoiding the fact that SkPaint cannot distinguish between various operations which have reduced bounds considerations (some operations are always strokes, others are always fills, some cannot produce a miter or an end cap, some may not be governed by a MaskFilter or PathEffect during rendering, etc). The SkPaint bounds computation method does not allow the caller to specify these conditions and so it would often produce bounds that were unnecessarily conservative because it took action based on attributes that didn't apply. Trying to "patch up" an SkPaint to reflect these conditions ended up even more complicated than just avoiding it in the first place.
- Avoiding some bugs in Skia bounds computations (which have all been reported). Most of these issues were due to a heavy reliance on developers only setting "the right attributes" that belonged with a given operation, but some of them resulted in under-computing the proper bounds (triggering a test failure in the canvas unit tests heavily updated in this PR). These issues have workarounds and you can see a number of lines of code in that file that refer to Skia bugs on a line that works around the referenced bugs...

For this PR the bounds accuracy testing in `display_list_canvas_unittests.cc` have been enabled by default and they now verify what should be reasonably optimal bounds representing a tradeoff between time and area. More importantly, there is an assert that our bounds are never larger than the bounds computed by SkPicture/Recorder for the tested cases which are fairly extensive, representing:

- Every drawing operation we support
- Rendered with every Paint attribute we support
- Rendered with a variety of Clip options
- Rendered with a variety of transforms
- Rendered with a few cases of save/saveLayer/restore

The matrix of tests is quite large and the bounds returned are both encompassing of all pixels that were rendered (or else a hard test failure results) and no looser than those computed by the Skia mechanism (which, for now, just reports the condition as it does not represent a functional program bug). (Note that our bounds accuracy performance is already being tracked by the recent changes to report cache metrics for all benchmarks to Skia Perf so we should get perf failures if our cache usage regresses.)

The tests were modified in a couple of key ways to try to verify these conditions:

- Some tests were rendering a simple version of the primitive that didn't exhibit any corner cases (such as path join miters in particular). The tests were modified where possible to try to trigger any conditions that are covered by conservative estimates in the bounds calculations. Basically, they now try to "earn" their conservative bounds.
- A  "tolerance" object was added to the bounds tests to help account for cases where the bounds would be calculated with a conservative consideration that the test case cannot correct for. For example, the bounds calculation code might compute the bounds of a simple horizontal line as if it has an unruly assortment of geometry that triggers worst cases for joins, end caps, and path effects. Unfortunately, simple lines can never do that and the code to detect that case in the bounds calculations is outside of our current concept of a tradeoff between accuracy and computational speed - so we add a tolerance to the testing computations in some of those cases to overlook or forgive that extra bounds padding.
- Another tolerance case is an SkTextBlob for which we rely on Skia to tell us what the bounds of the glyphs are - but they return a dramatic overestimate of those bounds based on maximal typographic metrics and obscure glyphs present in the font. Even using the supposedly homogenous `ahem` font still produces wildly over-estimated bounds for a simple string. So, a tolerance for the base bounds of a text blob are recorded.
- Finally, some of the adjustments are attribute-specific. To that end, an adjuster function is allowed which can dynamically update the tolerances based on the final rendering attributes. This is used to dynamically handle path effect tolerances for lines as well as targeting a relaxation of tolerances for some operations under a perspective transform where the bounds computed are highly conservative.

A minor, seemingly unrelated, change in the code was made to rename one of the parameters to drawShadow which had a name that conflicted with its usage. It shouldn't have been causing problems because it was just passed along from source to destination as "that flag that the developer handed us and which we need to remember and eventually pass to Skia". Since we never interpreted its value ourselves, the name didn't really matter, but I noticed it was wrong when I was updating the shadow tests.

### CODE REVIEW

**Comment by flar** [comment]
After re-running the internal memory consumption tests on this cherry-picked PR, it appears that this patch results in undoing the regressions of https://github.com/flutter/flutter/issues/88745

File: flow/display_list_utils.cc [hunk 1]
```diff
@@ -317,126 +431,158 @@ void DisplayListBoundsCalculator::drawPicture(const sk_sp<SkPicture> picture,
...
+  if ((flags & kIsAnyGeometryMask) != 0) {
+    if ((flags & kIsDrawnGeometry) != 0) {
+      FML_DCHECK((flags & (kIsFilledGeometry | kIsStrokedGeometry)) == 0);
+      flags |= style_flag_;
+    }
+
+    // Path effect occurs before stroking...
+    if (path_effect_) {
+      SkPathEffect::DashInfo info;
+      if (path_effect_->asADash(&info) == SkPathEffect::kDash_DashType) {
```
Comment (dnfield) [inline flow/display_list_utils.cc:482]: I don't believe we ever use this effect. There was a feature request for it that was long denied, and I think finally closed as won't fix. Users are expected to draw dashed lines manually.

File: flow/display_list_utils.h [hunk 2]
```diff
@@ -190,20 +190,32 @@ class SkMatrixDispatchHelper : public virtual Dispatcher,
...
+  ClipBoundsDispatchHelper() : ClipBoundsDispatchHelper(nullptr) {}
+
+  ClipBoundsDispatchHelper(const SkRect* cull_rect)
+      : has_clip_(cull_rect),
+        bounds_(cull_rect && !cull_rect->isEmpty() ? *cull_rect
+                                                   : SkRect::MakeEmpty()) {}
+
+  void clipRect(const SkRect& rect, bool is_aa, SkClipOp clip_op) override;
+  void clipRRect(const SkRRect& rrect, bool is_aa, SkClipOp clip_op) override;
+  void clipPath(const SkPath& path, bool is_aa, SkClipOp clip_op) override;
```
Comment (dnfield) [inline flow/display_list_utils.h:202]: I think the style guide would want these to be capitalized, but it looks like other things in this file are not, and we're not super consistent in the repo. 

That said, it'd probably be good for new code to prefer using PascalCase for method names in C++. But if you'd rather make that consistent in a separate PR that SGTM too.

File: flow/display_list_utils.h [hunk 3]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+      SkRect bounds = AccumulatorLayerData::getLayerBounds();
+      if (!getFilteredBounds(bounds, layer_filter_.get())) {
+        set_flooded();
+      }
+      return bounds;
+    }
 
    private:
-    static constexpr SkRect kMissingBounds = SkRect::MakeWH(-1, -1);
+    sk_sp<SkImageFilter> layer_filter_;
```
Comment (dnfield) [inline flow/display_list_utils.h:485]: Are we 100% sure that this will always get collected on the raster task runner?

File: flow/display_list_utils.h [hunk 4]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+
+  // A geometric operation that may be a stroke or fill operation
+  // depending on the current state of the paint Style attribute.
+  // This flag will automatically assume |kApplyMaskFilter|.
+  static constexpr int kIsDrawnGeometry = 0x04;
+
+  static constexpr int kIsAnyGeometryMask =  //
+      kIsFilledGeometry |                    //
+      kIsStrokedGeometry |                   //
+      kIsDrawnGeometry;
```
Comment (dnfield) [inline flow/display_list_utils.h:514]: Consider an `enum class`.

Comment (dnfield) [reply flow/display_list_utils.h:514]: (for this and below)

File: flow/display_list_utils.h [hunk 5]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+
+  static bool getFilteredBounds(SkRect& rect, SkImageFilter* filter);
+  bool adjustBoundsForPaint(SkRect& bounds, int flags);
+
+  void accumulateFlood();
+  void accumulateRect(const SkRect& rect, int flags) {
+    SkRect bounds = rect;
+    accumulateRect(bounds, flags);
+  }
+  void accumulateRect(SkRect& rect, int flags);
```
Comment (dnfield) [inline flow/display_list_utils.h:572]: Ditto on PascalCasing methods.

File: flow/display_list_utils.h [hunk 6]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+
+  // In very rare circumstances the ImageFilter is ignored.
+  // This is one of the few flags that turns off a step in
+  // estimating the bounds, rather than turning on any steps.
+  static constexpr int kIsUnfiltered = 0x40;
+
+  static constexpr SkScalar kMinStrokeWidth = 0.01;
+
+  skstd::optional<SkBlendMode> blend_mode_ = SkBlendMode::kSrcOver;
+  sk_sp<SkColorFilter> color_filter_;
```
Comment (dnfield) [inline flow/display_list_utils.h:550]: Is this safe to collect on non-raster task runners, and if not are we certain it will always get collected on the raster task runner?

File: flow/display_list_utils.h (hunk 3 above)
Comment (dnfield) [reply flow/display_list_utils.h:485]: (This one is definitely not safe to collect on non-raster task runners)

**Review by dnfield — COMMENTED** [review]
Cool! Mostly nits, main concern is around the threading questions.

File: flow/display_list_utils.cc (hunk 1 above)
Comment (flar) [reply flow/display_list_utils.cc:482]: Text uses it.

File: flow/display_list_utils.h (hunk 2 above)
Comment (flar) [reply flow/display_list_utils.h:202]: This is not new code in this PR. If we should change these, that should be done outside of this PR.

File: flow/display_list_utils.cc [hunk 7]
```diff
@@ -137,101 +137,210 @@ void SkMatrixDispatchHelper::reset() {
...
+    : ClipBoundsDispatchHelper(cull_rect) {
+  layer_infos_.emplace_back(std::make_unique<RootLayerData>());
+  accumulator_ = layer_infos_.back()->accumulatorForLayer();
+}
+void DisplayListBoundsCalculator::setCaps(SkPaint::Cap cap) {
+  cap_is_square_ = (cap == SkPaint::kSquare_Cap);
+}
+void DisplayListBoundsCalculator::setJoins(SkPaint::Join join) {
+  join_is_miter_ = (join == SkPaint::kMiter_Join);
+}
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.cc:215]: I capitalized them originally because any unclosed path has at least 2 caps and can have an arbitrary number of joins. I can change it, but the naming isn't really part of this PR.

File: flow/display_list_utils.h (hunk 4 above)
Comment (flar) [reply flow/display_list_utils.h:514]: These are mask constants...?

File: flow/display_list_utils.h (hunk 5 above)
Comment (flar) [reply flow/display_list_utils.h:572]: I went back and forth on this in the original DL PR. I can revisit this, but I'm not going to do it in this PR.

File: flow/display_list_utils.h (hunk 3 above)
Comment (flar) [reply flow/display_list_utils.h:485]: Since when? lib/ui/painting/image_filter.h holds them in an sk_sp and the SkPicture absolutely positively used to hold them in an sk_sp...?

File: flow/display_list_utils.cc (hunk 1 above)
Comment (dnfield) [reply flow/display_list_utils.cc:482]: Ahh haha I forgot that one. Ok.

File: flow/display_list_utils.h (hunk 6 above)
Comment (flar) [reply flow/display_list_utils.h:550]: Ditto to above. I can't find where we've ever held these other than in an sk_sp...?

File: flow/display_list_utils.h (hunk 4 above)
Comment (dnfield) [reply flow/display_list_utils.h:514]: Yes, you can still define them as an `enum class` to get some more type safety, but I believe you'd have to overload any operators you use to do so. Your call.

File: flow/display_list_utils.h (hunk 3 above)
Comment (dnfield) [reply flow/display_list_utils.h:485]: Are we just lucking out then? An Image filter will almost always have an SkImage inside, and that isn't safe to destroy on the UI task runner, which is why we use a SkiaGPUObject. There are race conditions involved though, so maybe it just works out most of the time?

/cc @jason-simmons or @chinmaygarde who might know something I'm missing here.

**Comment by flar** [comment]
I filed https://github.com/flutter/flutter/issues/90232 and linked it into https://github.com/flutter/flutter/issues/85737 for updating the DisplayList API names.

File: flow/display_list_utils.h (hunk 6 above)
Comment (dnfield) [reply flow/display_list_utils.h:550]: I suspect the `SkColorFilter` is safe to release on a non-GPU thread, since I don't think it ever ends up holding any GPU related resources.

File: flow/display_list_utils.h (hunk 3 above)
Comment (jason-simmons) [reply flow/display_list_utils.h:485]: yes - it looks like `flutter::ImageFilter` and `flutter::ImageFilterLayer` should be wrapping the `SkImageFilter` in a `SkiaGPUObject`.

`SkImageFilter` implementations (specifically `SkImageImageFilter`) may hold references to `SkImage` objects.  So these objects should be dereferenced on the raster thread.

Comment (flar) [reply flow/display_list_utils.h:485]: I'd like to hear more about this because I think we have a long history of storing these as just regular shared pointers.

On a related front, it looks like we store images as `flutter::SkiaGPUObject<SkImage>` in the lib/ui/painting/image.cc class, but when we add them to a display list or an SkPicture we extract the Skia object and just hold onto that instead. That's probably where there may be a bug lurking if anywhere...?

Comment (flar) [reply flow/display_list_utils.h:485]: Flutter doesn't expose any ImageFilter that holds a reference to an image. We have a native method to do so in the engine, but I don't see any calls to it from the framework...?

Comment (flar) [reply flow/display_list_utils.h:485]: Ah, the Dart Picture native wrapper holds the SkPicture in a `SkiaGPUObject`, but the SkPicture it holds can outlive that native object, so that seems to be an attempt to solve this problem, but I don't see how it is robust.

We currently only hold the DisplayList  in an sk_sp. We can upgrade that to a SkiaGPUObject. I'll file a bug on it.

Comment (flar) [reply flow/display_list_utils.h:485]: https://github.com/flutter/flutter/issues/90237

Comment (flar) [reply flow/display_list_utils.h:485]: How does our use of `SkPicture` work where you can embed an `SkImage` using an `sk_sp`, and elsewhere the `SkiaGPUObject` from whence it came could be reclaimed, but the `sk_sp` version is still referenced inside the `SkPicture`?

Comment (flar) [reply flow/display_list_utils.h:485]: Where do `ImageImageFilter`s come from anyway? The framework doesn't seem to have a method to make them.

Comment (jason-simmons) [reply flow/display_list_utils.h:485]: Interesting - the `flutter::ImageFilter` native peer class defines a method for constructing an `SkImageImageFilter`

But it looks like the engine Dart APIs do not actually expose that method to Dart code.

If `ImageFilter::initImage` is unused then it should be removed.  And if none of the `SkImageFilter` subclasses that the engine creates are holding on to any GPU resources, then the lack of the `SkiaGPUObject` wrapper is not an issue.

`SkiaGPUObject` is needed for objects that will be deleted on the UI thread but are referencing other objects that must be deleted on the raster thread.  It's used for Dart native peer objects that hold references to Skia objects whose destructors may call graphics APIs.

If an `sk_sp<SkPicture>` is guaranteed to be deleted on the raster thread then it does not need the `SkiaGPUObject` wrapper.

File: flow/display_list_utils.h (hunk 4 above)
Comment (chinmaygarde) [reply flow/display_list_utils.h:514]: I like the idea of using an enum if we don't want to make it an `enum class` which would need overloading the operators. Just groups the various values better IMO.

```
enum FlagValue {
  kSomething        = 1 << 0,
  kSomethingElse = 1 << 1,
  kAnotherThing   = 1 << 2,
};
```

File: flow/display_list_utils.h (hunk 6 above)
Comment (chinmaygarde) [reply flow/display_list_utils.h:550]: `SkColorFilters` are `SkFlattenable`'s. If they held GPU resources, they could be flattened. So I think this is safe. They are also [documented to be shareable across threads](https://github.com/google/skia/blob/941812fdcc8426b29aca8072326cbeb9371cc9a3/include/core/SkColorFilter.h#L25).

File: flow/display_list_utils.h (hunk 3 above)
Comment (flar) [reply flow/display_list_utils.h:485]: `sk_sp<SkPicture>` is not guaranteed to be deleted on the raster thread, it is a regular shared pointer.

`SkiaGPUObject<SkPicture>` is guaranteed to be deleted on the raster thread. `flutter::Picture` holds a `SkiaGPUObject<SkPicture>`. `flutter::PictureLayer` also holds a copy via `SkiaGPUObject`. If either of them hand out an extra copy of the skia version of the picture, then we need to hope that that reference gets unrefed before either of those objects are unrefed.

In either case, it's not clear that `SkPicture` itself needs to be deleted on the raster thread unless it holds a pointer to a Skia object that also needs to be deleted on the raster thread (SkImage or anything that holds an SkImage - anything else?).

But what about an `SkImage` that is held by an `SkPicture`? If the `flutter::Image` goes away before the `flutter::Picture` then its `SkiaGPUObject` will deref and potentially queue the `SkImage` for deletion. When that `SkiaGPUObject` reaches the raster thread, it will get deref'd, but it won't get freed! Instead, the skia object itself will be held by the `SkPicture`. Fortunately, if we also delete the SkPicture on the raster thread, then it should work out.

For now, we could add a `SkiaGPUObject` wrapper for DisplayList objects in `flutter::Picture` and `flutter::PictureLayer`. I submitted an Issue for that (mentioned above: https://github.com/flutter/flutter/issues/90237) and will take care of that separately.

**Review by chinmaygarde — COMMENTED** [review]
Thanks for the great PR commit message. Helped me understand what you were going for. I took a close look at the unit-tests but must admit I could not follow how the tolerances were setup. Some other minor comments but otherwise looks good to me.

File: flow/display_list_utils.h (hunk 6 above)
Comment (flar) [reply flow/display_list_utils.h:550]: I think so too. And if we get rid of the vestigial code that attempts to create an `SkImageImageFilter` then the same could be said for our `SkImageFilter` uses.

File: flow/display_list_utils.h (hunk 3 above)
Comment (flar) [reply flow/display_list_utils.h:485]: Is there a reason that the GPU resources in the SkImage object can't be taught to manage this themselves? This seems like a house of cards built around an object that might potentially hold thread-constrained resources causing anyone who might touch one to have to jump through hoops. The GPU resource itself should protect itself without any outside burden...

Comment (jason-simmons) [reply flow/display_list_utils.h:485]: The `SkiaGPUObject<SkPicture>` itself will be deleted on the UI thread.  `SkiaGPUObject<SkPicture>` wraps an `sk_sp<SkPicture>`, and `SkiaGPUObject's` destructor will queue deletion of the `sk_sp<SkPicture>` to the raster thread.

For a specific `sk_sp<SkPicture>` instance it may be the case that it is used exclusively by code running on the raster thread.  If so, then that `sk_sp<SkPicture>` will be destructed safely on the raster thread and `SkiaGPUObject` is therefore unnecessary.

If there is a chance that an `sk_sp<SkPIcture>` instance might be destructed on the UI thread, then it must be wrapped with a `SkiaGPUObject` so that any resources such as `SkImages` referenced by the `SkPicture` will be deleted on the raster thread.

The same would apply to a `DisplayList` if it could be destructed on the UI thread.

File: flow/display_list_utils.h (hunk 4 above)
Comment (flar) [reply flow/display_list_utils.h:514]: I see lots of tutorials on this, but none of them answer the basic question of "why".

enums are enumerable. You can see all of its values right there.

Except if it is flags then you don't see all of its values right there. You can make up other values by combining them. Which of the enumerated values is `kIsFrog | kIsZebra`. I don't see a "zebra frog" in the list and such an animal would not likely survive long.

So, this concept has philosophical problems.

What advantages does it provide? "Hey, it's an enum!" - and how does that help anything?

One of the advantages of enums is knowing that your switch statements are complete, but you don't use flags in switches and if you did this aspect of enums would lie to you. As long as you had a `case` for each bit flag (which would be a horribly inefficient switch table) then it would think things were complete.

I suppose it has a type associated with it, but I feel like it is expressing "apples" as if they were "oranges".

File: flow/display_list_utils.h (hunk 3 above)
Comment (jason-simmons) [reply flow/display_list_utils.h:485]: Skia considers it the application's responsibility to configure threads to access any required graphics APIs.

On a platform such as OpenGL/EGL, the raster thread has a context for calling OpenGL APIs but the UI thread does not.  So the engine needs to ensure that any code that might invoke those APIs (such as Skia object destructors) runs on the raster thread.

File: flow/display_list.cc [hunk 8]
```diff
@@ -1365,10 +1365,10 @@ void DisplayListBuilder::drawTextBlob(const sk_sp<SkTextBlob> blob,
 void DisplayListBuilder::drawShadow(const SkPath& path,
                                     const SkColor color,
                                     const SkScalar elevation,
-                                    bool occludes,
+                                    bool transparentOccluder,
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list.cc:1368]: yeah, since I'm introducing this here I should get the name right. Still, I don't like "transparent_occluder". Why? Because the name is a not-name, or an oxymoron, or a self-contradicting name and it is why I ran into this in the first place. "this is a flag about occlusion!" "OK, so it hides something, right? That's what occlusion refers to." "No, um, this flag is about failing to occlude something." "But it has 'occlude' right there in its name???"

It would be less confusing if it had a "not" word in it, but you have to understand what transparency is and what occlusion is to notice that they contradict each other and so this is a flag about something that isn't there.

The record name was "NonOccludingOp". I'm trying to think of a better word.

Option B, reverse its meaning in DL, but then it would be negating its value going into and out of the DL stream and any code that was dispatching or being dispatched to would have to recognize that this usage is negated from the Skia usage.

File: flow/display_list_canvas_unittests.cc [hunk 9]
```diff
@@ -45,6 +47,200 @@ constexpr SkRect TestBounds = SkRect::MakeWH(TestWidth, TestHeight);
...
     SkRect::MakeLTRB(RenderLeft, RenderTop, RenderRight, RenderBottom);
 
+// The tests try 3 miter limit values, 0.0, 4.0 (the default), and 10.0
+// These values will allow us to construct a diamond that spans the
+// width or height of the render box and still show the miter for 4.0
+// and 10.0
+
+// The X offsets which will be used for tall vertical diamonds are
+// expressed in terms of the rendering height to obtain the proper angle
+constexpr SkScalar MiterExtremeDiamondOffsetX = RenderHeight * 0.04;
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_canvas_unittests.cc:57]: I'll add some text about how I experimented with Skia fiddle to see how close I could get the miters before they were cut off at various limits.

File: flow/display_list_utils.h (hunk 3 above)
Comment (dnfield) [reply flow/display_list_utils.h:485]: It looks like SkImageFilter is an SkFlattenable, so I guess it's ok? A quick look at the headers related to it don't seem to show it retaining any GPU resources.

File: flow/display_list_canvas_unittests.cc [hunk 10]
```diff
@@ -136,51 +368,126 @@ class CanvasCompareTester {
...
+            [=](DisplayListBuilder& b) {
+              b.setImageFilter(filter);
+              b.saveLayer(nullptr, true);
+              b.setImageFilter(nullptr);
+              b.setStrokeWidth(5.0);
+            },
+            cv_restored, dl_restored, adjuster, blur5Tolerance,
+            "saveLayer ImageFilter, no bounds");
+      }
+      ASSERT_TRUE(filter->unique()) << "SL with IF no bounds Cleanup";
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_canvas_unittests.cc:439]: :( then the lines will wrap...

File: flow/display_list_utils.h (hunk 3 above)
Comment (dnfield) [reply flow/display_list_utils.h:485]: And @flar - we have `PictureLayer`s create their own SkiaGPUObject wrapping the SkPicture for that reason.

File: flow/display_list_utils.h (hunk 4 above)
Comment (dnfield) [reply flow/display_list_utils.h:514]: If you use an enum class the compiler will help make sure you use the right types and don't end up accidentally mixing things up.  But it does require operator overloading

File: flow/display_list_canvas_unittests.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_canvas_unittests.cc:439]: I fixed all of the abbreviations in the info strings I could find.

File: flow/display_list_utils.h [hunk 11]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+    //
+    // Flutter is unlikely to ever run into this as the Dart
+    // mechanisms all supply a cull rect for all Dart Picture
+    // objects, even if that cull rect is kGiantRect.
+    void set_flooded() { is_flooded_ = true; }
+
+    // |is_flooded| should be called after |getLayerBounds| in case
+    // a problem was found during the computation of those bounds,
+    // the layer will have one last chance to flag a flooded state.
+    bool is_flooded() { return is_flooded_; }
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:409]: Check

File: flow/display_list_utils.h [hunk 12]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
-   protected:
-    BoundsAccumulator* saved_accumulator_;
+    LayerData(BoundsAccumulator* outer) : outer_(outer), is_flooded_(false) {}
+    virtual ~LayerData() = default;
+
+    virtual BoundsAccumulator* accumulatorForLayer() { return outer_; }
+    virtual BoundsAccumulator* accumulatorForRestore() { return outer_; }
+    virtual SkRect getLayerBounds() { return SkRect::MakeEmpty(); }
+
+    // is_flooded is to true if we ever encounter an operation
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:380]: Only if they failed to provide a cull rect or surround these operations with a clip.

I'm following what happens with SkPicture. If you record one with no bounds then the cull rect defaults to an empty rect. When they encounter an op with no bounds (drawColor for instance) then they accumulate the cull rect - which is empty if it was defaulted - which is a NOP.

Essentially I do the same thing, but at least I set a flag indicating that you had an unbounded operation with no cull rect or clip to contain it.

If I returned an infinite bounds then you'd lose any information on what was drawn that was reasonably bounded. With Skia you'd at least get the bounds of the finite operations and so I wanted to follow suit.

File: flow/display_list_utils.h [hunk 13]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
-    virtual BoundsAccumulator* restore();
-
-   protected:
-    BoundsAccumulator* saved_accumulator_;
+    LayerData(BoundsAccumulator* outer) : outer_(outer), is_flooded_(false) {}
+    virtual ~LayerData() = default;
+
+    virtual BoundsAccumulator* accumulatorForLayer() { return outer_; }
+    virtual BoundsAccumulator* accumulatorForRestore() { return outer_; }
+    virtual SkRect getLayerBounds() { return SkRect::MakeEmpty(); }
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:378]: If they were pure virtual then I'd need another class just to implement the most common trivial implementations.

File: flow/display_list_utils.h [hunk 14]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+    bool is_flooded_;
+
+    FML_DISALLOW_COPY_AND_ASSIGN(LayerData);
   };
 
-  class SaveLayerInfo : public SaveInfo {
+  // An intermediate implementation class that handles keeping
+  // a local accumulator for the layer, used by both |RootLayerData|
+  // and |SaveLayerData|.
+  class AccumulatorLayerData : public LayerData {
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:421]: AccumulatorLayerData can't be final. It's got a subclass (2 of them actually). I suppose I could finalize the leaf classes.

File: flow/display_list_utils.h [hunk 15]
```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
...
+   private:
+    FML_DISALLOW_COPY_AND_ASSIGN(RootLayerData);
+  };
 
-   protected:
-    DisplayListBoundsCalculator* calculator_;
+  // Used for |save|
+  class SaveData : public LayerData {
+   public:
+    using LayerData::LayerData;
```
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:456]: Unsure why what is necessary?

File: flow/display_list_utils.h (hunk 3 above)
Comment (flar) [reply flow/display_list_utils.h:516]: For the short term I've fixed this under https://github.com/flutter/engine/pull/28679

Eventually we will overhaul how we deal with Skia objects inside the DisplayList and we may be able to relax this restriction on holding DisplayList objects in GPU object wrappers. For now, that PR should make DisplayList as robust as SkPicture is currently...

File: flow/display_list_canvas_unittests.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_canvas_unittests.cc:60]: New comments are added now.

File: flow/display_list_utils.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h]: I also changed the naming from "flooded" to "unbounded". I still use the word flood in a couple of places to describe the behavior of drawColor/drawPaint - they flood the surface. But the results are now referred to as "unbounded" wrt the Layer structures and the Calculator's external API.

File: flow/display_list_utils.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:450]: I made the leaf versions final.

**Comment by flar** [comment]
Still TBD - a better name for "transparentOccluder" that doesn't parse as "the thing, except its not the thing because it has a flaw".

I am not planning on switching the flags to enums at this point.

Naming style issues will be considered in a sweep once the functional bugs are all fixed. See https://github.com/flutter/flutter/issues/90232

File: flow/display_list_canvas_unittests.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply flow/display_list_canvas_unittests.cc:60]: Ok, I was trying to work them out. Eh, magic numbers are magic.

File: flow/display_list_utils.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (chinmaygarde) [reply flow/display_list_utils.h:485]: ~`using LayerData::LayerData`?~

Discussed offline. This is for passthrough ctors. I usually just declare another ctor in the subclass that calls the appropriate base class ctor. Just unfamiliar with the pattern.

**Review by chinmaygarde — APPROVED** [review]
Looks good to me. Are there any open questions from other reviewers to be handled here?

File: flow/display_list_utils.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.h:485]: https://en.cppreference.com/w/cpp/language/using_declaration#Inheriting_constructors

**Comment by flar** [comment]
> Looks good to me. Are there any open questions from other reviewers to be handled here?

@dnfield ?

File: flow/display_list_utils.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list_utils.cc:215]: I added a note in the "fix up DL naming" Issue about these names.

**Comment by flar** [comment]
It looks like dnfield cannot get back to this today. I don't think he expressed any functional concerns (other than the GPU lifetime concerns that were already patched by another PR) and I believe that the code style concerns he expressed can be dealt with under flutter/flutter#90232 so I'm going to land this and can address any remaining concerns with future PRs.

File: flow/display_list_utils.h (hunk 2 above)
Comment (flar) [reply flow/display_list_utils.h:202]: In revisiting this in https://github.com/flutter/engine/pull/28753 I decided to leave these specific method names consistent with the names in SkCanvas and SkPaint that they are replacing, but I looked to make method names consistent with the style guidelines. Please follow up in that PR with further discussion, but I wanted to make a note here to remind myself of where further discussion of these suggestions went.

File: flow/display_list_utils.h (hunk 3 above)
Comment (flar) [reply flow/display_list_utils.h:516]: Just to follow up here - after a discussion with Skia it looks like the intention is for all types to be thread-safe with respect to construction and deletion and they all but promise that for any type that can be held in an SkPaint, except that it is possible to run afoul of the one area of the API that can't (yet) make that promise - SkImage objects. They also largely all but make that promise for SkPicture except that it can also hold references to SkImage objects.

Fixing the SkImage issue was a planned goal for the next major revision, but they acknowledge that it is creating technical debt for some of their customers, including Flutter, so they are looking for opportunities for moving that time frame up. No promises on this front, but I wanted to document here what we know about the situation since there was speculation in this comment thread.

File: flow/display_list_utils.h (hunk 5 above)
Comment (flar) [reply flow/display_list_utils.h]: Many of these got either PascalCased or accessor_formatted in https://github.com/flutter/engine/pull/28753 per the style guide.

File: flow/display_list.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (flar) [reply flow/display_list.cc:1368]: As mentioned in https://github.com/flutter/engine/pull/28753 I left it as transparent_occluder which at least follow our style guide on capitalization of parameters, but it leaves the questionable name in play. Since that naming is used all the way from Flutter classes through the PhysicalShapeLayer through to Skia, I decided to leave it in the DisplayList code, but with corrected format `transparent_occluder`.
